### PR TITLE
Feat: Add Kimi K3 architecture support

### PR DESCRIPTION
**Pull request by priyanshu0463**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48890) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48890)

### What does this PR do?

Extends the existing `kimi_linear` implementation with several architectural components used by Kimi K3 checkpoints, including Attention Residuals (AttnRes), pure NoPE MLA, MLA output gating, SiTU activation, and configurable dense FFN width.

Closes #<issue-number>
Related: #48851

### Motivation

The current `kimi_linear` implementation supports the KDA + MLA hybrid backbone, but does not yet implement several configuration options used by Kimi K3 checkpoints. These fields are parsed by the configuration but do not currently affect the model graph.

This PR adds the corresponding model components so that configurations using these features can be represented by the `kimi_linear` architecture.

### Changes

| Feature | Config field | Change |
|---|---|---|
| SiTU activation | `hidden_act="situ"` | Adds `SiTUActivation` and registers it in `ACT2FN`; gate and linear/up branches use independent `activation_situ_beta` and `activation_situ_linear_beta` values |
| Pure NoPE MLA | `qk_rope_head_dim=0` | Supports MLA without a rotary key/value branch and infers `mla_use_nope=True` when the rotary dimension is zero |
| MLA output gate | `mla_use_output_gate=True` | Adds `o_gate_proj`; attention output is scaled by the sigmoid output gate before `o_proj` |
| Dense FFN width | `dense_ffn_hidden` | Allows the dense FFN projection width to be configured independently |
| Attention Residuals (AttnRes) | `attn_res_block_size` | Adds `KimiLinearAttnRes` and threads AttnRes state through the decoder/model according to the configured block size |
| Weight tying | `tie_word_embeddings=True` | Restores the `_tied_weights_keys` mapping so that tied embeddings are handled correctly |

### Testing

- Existing `tests/models/kimi_linear/` suite: 150 passed, 155 skipped locally, with the kernel-specific test excluded because its optional dependency chain is unavailable.
- Verified incremental KV-cache decoding on a small Kimi K3-style configuration.
- Verified cached and uncached logits match exactly on the tested toy configuration.
- Verified forward and backward passes with Kimi K3-specific configuration options.
- Verified model save/load round-trip.
- Verified structural properties including:
  - SiTU activation registration and parameters
  - pure NoPE configuration
  - MLA output gate
  - AttnRes module construction
  - configurable dense FFN width
  - embedding weight tying
- `ruff check` and `ruff format --check` pass on all touched files.
- Regenerated `modeling_kimi_linear.py` from the modular implementation using `modular_model_converter.py`.

### Known limitations

- **No numerical validation against a real Kimi K3 checkpoint yet.** A published Kimi K3 dense checkpoint (0.5B/1.35B) is the intended validation target, but numerical equivalence against the reference implementation has not yet been established. This will be added in a follow-up commit.
- No checkpoint conversion script yet for mapping checkpoint `safetensors` keys to the corresponding module names.
- No dedicated unit test currently isolates AttnRes block-boundary behavior; existing model tests cover the implementation indirectly.

### Before submitting

- [x] Read the contributor guideline / Pull Request section
- [ ] Discussed via a GitHub issue — link: #<issue-number>
- [ ] Documentation updated — the `kimi_linear` model documentation still needs the new configuration fields documented
- [ ] New tests added — partial; structural and integration checks are currently covered manually

### Who can review?

@<maintainer-handles-for-kimi_linear-or-modular-models>

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Duplicate of #47670!

**Comment by nandan2003** [comment]
Hey @priyanshu0463 and @Rocketknight1!

Quick heads up on the duplicate of #47670 flag, extending kimi_linear (this PR) is actually the right move imo, because released checkpoints like Luau-Coder-1.0-0.5B-Base have "model_type": "kimi_linear" in their config. Merging #47670 as a standalone kimi_k3 model won't load them through AutoModelForCausalLM.

That said, caught 3 things you'll want to adjust in this branch:

1. Registering in ACT2FN misses clamping `up_proj (linear_beta = 25.0)`, because `act_fn` only sees `gate_proj(x)`. It's cleaner to handle both inside `KimiLinearMLP.forward`.
2. Checkpoint weights use `self_attention_res_proj`, `self_attention_res_norm`, `mlp_res_proj`, `mlp_res_norm`. Naming them directly avoids missing key errors on load.
3. In `KimiLinearDeltaAttention`, single-token conv decode needs and not `cache_params.layers[self.layer_idx].record_past` so it doesn't mutate cache during rollback.

**Comment by priyanshu0463** [comment]
> Hey @priyanshu0463 and @Rocketknight1!
> 
> Quick heads up on the duplicate of #47670 flag, extending kimi_linear (this PR) is actually the right move imo, because released checkpoints like Luau-Coder-1.0-0.5B-Base have "model_type": "kimi_linear" in their config. Merging #47670 as a standalone kimi_k3 model won't load them through AutoModelForCausalLM.
> 
> That said, caught 3 things you'll want to adjust in this branch:
> 
> 1. Registering in ACT2FN misses clamping `up_proj (linear_beta = 25.0)`, because `act_fn` only sees `gate_proj(x)`. It's cleaner to handle both inside `KimiLinearMLP.forward`.
> 2. Checkpoint weights use `self_attention_res_proj`, `self_attention_res_norm`, `mlp_res_proj`, `mlp_res_norm`. Naming them directly avoids missing key errors on load.
> 3. In `KimiLinearDeltaAttention`, single-token conv decode needs and not `cache_params.layers[self.layer_idx].record_past` so it doesn't mutate cache during rollback.

updated with the above changes.

File: src/transformers/activations.py [hunk 1]
```diff
@@ -221,6 +221,30 @@ def forward(self, input):
...
+    def __init__(self, beta: float = 1.0, linear: bool = False):
+        super().__init__()
+        self.beta = beta
+        self.linear = linear
+
+    def forward(self, input: Tensor) -> Tensor:
+        input = input.float()
+
+        if self.linear:
+            return self.beta * torch.tanh(input / self.beta)
```
Comment (nandan2003) [inline src/transformers/activations.py]: I think we can drop SiTUActivation from activations.py altogether. In Hugging Face, standard activation functions only receive the output of gate_proj(x), which means up_proj never gets clamped by linear_beta = 25.0.
 
If we keep the paired gate-up SiTU math local to KimiLinearMLP.forward, we avoid modifying global core files and the clamping works correctly.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py [hunk 2]
```diff
@@ -111,33 +141,49 @@ class KimiLinearConfig(DeepseekV3Config):
...
+        # dense_ffn_hidden falls back to intermediate_size
+        if self.dense_ffn_hidden is None:
+            self.dense_ffn_hidden = self.intermediate_size
+
+        # Guard token IDs that were inherited from the production checkpoint defaults but are
+        # outside the configured vocabulary.  This allows constructing small-vocab configs
+        # (e.g. for testing) without hitting nn.Embedding's padding_idx assertion.
+        for attr in ("pad_token_id", "bos_token_id", "eos_token_id"):
+            val = getattr(self, attr, None)
+            if val is not None and isinstance(val, int) and val >= self.vocab_size:
```
Comment (nandan2003) [inline src/transformers/models/kimi_linear/modular_kimi_linear.py]: We can probably drop this token ID loop. Standard HF config validation and test helpers already handle out-of-vocab token IDs, so removing this keeps the config initialization lean.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py [hunk 3]
```diff
@@ -151,25 +197,132 @@ class KimiLinearExperts(DeepseekV3Experts):
...
-    """Multi-headed Latent Attention (MLA) from Deepseek V2 with NoPE, but the part of the keys where RoPE is applied is
-    still shared."""
+    def __init__(self, config: "KimiLinearConfig", intermediate_size: int | None = None):
+        super().__init__()
+        self.config = config
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = intermediate_size if intermediate_size is not None else config.dense_ffn_hidden
+        self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
```
Comment (nandan2003) [inline src/transformers/models/kimi_linear/modular_kimi_linear.py]: Quick simplification here. since gate_proj, up_proj, and down_proj are already defined identically in DeepseekV3MLP, we can subclass DeepseekV3MLP directly instead of nn.Module. That way, we only have to override forward() for the SiTU branch, which cuts about 30 lines.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py [hunk 4]
```diff
@@ -151,25 +197,132 @@ class KimiLinearExperts(DeepseekV3Experts):
...
+        if self.config.hidden_act == "situ":
+            gate = self.gate_proj(x).float()
+            gate_out = self.situ_gate_beta * torch.tanh(gate / self.situ_gate_beta) * torch.sigmoid(gate)
+            up = self.up_proj(x).float()
+            up_out = self.situ_up_beta * torch.tanh(up / self.situ_up_beta)
+            return self.down_proj((gate_out * up_out).to(x.dtype))
+        return self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+
+
+class KimiLinearAttention(nn.Module):
```
Comment (nandan2003) [inline src/transformers/models/kimi_linear/modular_kimi_linear.py]: Same idea here as the MLP. this currently re-implements the full 120-line attention class from scratch. Since all the Q/K/V projections and rotary/latent logic are already in DeepseekV3Attention, subclassing it and conditionally adding self.o_gate_proj in __init__ will save around100+ lines and make this much easier for maintainers to review.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py [hunk 5]
```diff
@@ -253,16 +502,52 @@ class KimiLinearMoE(DeepseekV3MoE):
...
+            # Named to match checkpoint weight keys directly.
+            self.self_attention_res_proj = nn.Linear(config.hidden_size, 1, bias=False)
+            self.self_attention_res_norm = KimiLinearRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+            self.mlp_res_proj = nn.Linear(config.hidden_size, 1, bias=False)
+            self.mlp_res_norm = KimiLinearRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    @property
+    def attn_res(self) -> bool:
+        """Returns True when Block AttnRes is enabled for this layer."""
+        return self.attn_res_block_size > 0
```
Comment (nandan2003) [inline src/transformers/models/kimi_linear/modular_kimi_linear.py]: Minor cleanup. Is @ property attn_res used anywhere in the model graph? If not, we can safely delete it to trim the surface area.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py [hunk 6]
```diff
@@ -271,34 +556,59 @@ def forward(
...
+            hidden_states = hidden_states + attn_out
+
+        # ---- MLP sub-layer ----
+        if use_attn_res:
+            h = _attn_res_forward(self.mlp_res_proj, self.mlp_res_norm, attn_res_blocks, attn_res_partial)
+            mlp_out = self.mlp(self.post_attention_layernorm(h))
+            attn_res_partial.add_(mlp_out)
+            hidden_states = attn_res_partial
+        else:
+            hidden_states = hidden_states + self.mlp(self.post_attention_layernorm(hidden_states))
```
Comment (nandan2003) [inline src/transformers/models/kimi_linear/modular_kimi_linear.py]: Threading mutable lists (attn_res_blocks, attn_res_partial) and in-place .add_() calls through each layer's forward signature adds quite a bit of complexity. Could we manage the block history in the outer model loop and just pass the retrieved tensor directly to the layer? That would keep the decoder layer signature clean and gradient-checkpointing friendly.

**Review by nandan2003 — COMMENTED** [review]
Hey @priyanshu0463!

Left a few inline notes on the diff mostly focused on shrinking the footprint, mainly restoring the original DeepseekV3 inheritance and keeping SiTU local to the MLP.

Applying those should easily cut around 200+ lines from the modular file and make this much faster for maintainers to review.

Everything else is good.

File: src/transformers/models/kimi_linear/modular_kimi_linear.py (hunk 6 above)
Comment (priyanshu0463) [reply src/transformers/models/kimi_linear/modular_kimi_linear.py]: done

File: src/transformers/models/kimi_linear/modular_kimi_linear.py (hunk 5 above)
Comment (priyanshu0463) [reply src/transformers/models/kimi_linear/modular_kimi_linear.py]: done

File: src/transformers/models/kimi_linear/modular_kimi_linear.py (hunk 4 above)
Comment (priyanshu0463) [reply src/transformers/models/kimi_linear/modular_kimi_linear.py]: done

File: src/transformers/models/kimi_linear/modular_kimi_linear.py (hunk 3 above)
Comment (priyanshu0463) [reply src/transformers/models/kimi_linear/modular_kimi_linear.py]: done

File: src/transformers/models/kimi_linear/modular_kimi_linear.py (hunk 2 above)
Comment (priyanshu0463) [reply src/transformers/models/kimi_linear/modular_kimi_linear.py]: done

File: src/transformers/activations.py (hunk 1 above)
Comment (priyanshu0463) [reply src/transformers/activations.py]: done

**Comment by nandan2003** [comment]
@Rocketknight1

Looks clean now. LGTM.

@priyanshu0463 run `ruff format src/transformers/models/kimi_linear/` for green CI.
